activation/certifier.go

```diff
@@ -90,139 +93,67 @@ func NewCertifier(
 	client certifierClient,
 ) *Certifier {
 	c := &Certifier{
-		client: client,
-		logger: logger,
-		db:     db,
+		client:         client,
+		logger:         logger,
+		db:             db,
+		certifications: make(map[string]chan struct{}),
 	}
 
 	return c
 }
 
-func (c *Certifier) Certificate(id types.NodeID, poet string) *certifier.PoetCert {
-	cert, err := certifier.Certificate(c.db, id, poet)
+func (c *Certifier) Certificate(
+	ctx context.Context,
+	id types.NodeID,
+	certifier *url.URL,
+	pubkey []byte,
+) (*certifierdb.PoetCert, error) {
+	// We index certs in DB by node ID and pubkey. To avoid redundant queries, we allow only 1
+	// request per (nodeID, pubkey) pair to be in flight at a time.
+	key := string(append(id.Bytes(), pubkey...))
+	c.certificationsLock.Lock()
+	if ch, ok := c.certifications[key]; ok {
+		c.certificationsLock.Unlock()
+		<-ch
+	} else {
+		ch := make(chan struct{})
+		c.certifications[key] = ch
+		c.certificationsLock.Unlock()
+		defer func() {
+			c.certificationsLock.Lock()
+			close(ch)
+			delete(c.certifications, key)
+			c.certificationsLock.Unlock()
+		}()
+	}
```

Comment:
Take a look at the `singleflight` package: https://pkg.go.dev/golang.org/x/sync/singleflight

Maybe it is worth considering using that over implementing something similar ourselves - we already have a dependency on the `golang.org/x/sync` package

Reply:
Changed to use this :+1: